ZenProfiles and centralized ZenML repositories

stefannica opened the pull request:
## ZenProfiles and centralized ZenML repositories

This PR adds a new concept to ZenML that creates an additional
dimension to how users can configure stacks for their projects: *Profiles*.
The Profile replaces the Repository as the container for Stacks. 

### Quick Intro

The first time ZenML users run a CLI command on their machine (or any
ZenML code, for that matter) that requires access to a Repository, ZenML
initializes a default Profile and the default Stack:

![Screenshot from 2022-03-21 09-43-13](https://user-images.githubusercontent.com/3963946/159229075-6f9b073c-cfa0-4660-9792-d43d254a8727.png)

Initializing a ZenML Repository is optional. Users can configure stacks
and they will be stored into the active profile. The active stack is also
configured in the active profile:

![Screenshot from 2022-03-21 09-50-15](https://user-images.githubusercontent.com/3963946/159229549-0235d5f6-2253-4296-986d-4ab2697dfbbd.png)

Initializing a Repository root is still supported (and recommended), but
it is no longer where stacks are persisted. An initialized repository is
just a way of configuring the active profile and active stack at directory level:

![Screenshot from 2022-03-21 09-56-10](https://user-images.githubusercontent.com/3963946/159230094-578f00cf-5c54-45c6-9144-f4f24a3b0bd4.png)

New profiles can be created and they are automatically initialized with
the default stack configuration:

![Screenshot from 2022-03-21 09-59-18](https://user-images.githubusercontent.com/3963946/159230527-4d2192ad-1fbf-4c69-a054-2982d64f5cf2.png)

Legacy .zen folders that contain stacks configurations are detected
and the user is given a choice to migrate them to a profile:

![Screenshot from 2022-03-21 10-23-11](https://user-images.githubusercontent.com/3963946/159233813-86fb3863-a8cb-4abc-88e4-4dd866602925.png)

### Implementation Details

* _GlobalConfig_ and _Repository_ are refactored into singletons via
metaclasses. Calling GlobalConfig() and Repository() multiple times
returns the same instance. This is required to guarantee a single point
of entry for read-write access to the store where information is persisted.

* multiple _ConfigProfile_ instances are now managed under the _GlobalConfig_.
This is an example of what the global configuration file looks like:

```yaml
activated_profile: stefan
analytics_opt_in: true
profiles:
  default:
    active_stack: default
    name: default
    store_type: local
    store_url: file:///home/stefan/.config/zenml/profiles/default
  stefan:
    active_stack: default
    name: stefan
    store_type: local
    store_url: file:///home/stefan/.config/zenml/profiles/stefan
user_id: fb1c5a55-31b5-4e73-8d3f-dc2430ac26d7
version: 0.6.3
```

* initializing repositories with `zenml init` is no longer mandatory. The active
profile is now initialized automatically in the global configuration location
(e.g. `~/.config/zenml` on Linux) and can be used to manage stacks globally.

* `zenml init` can still be used to initialize repository directories to be used
as repository roots (i.e. local `.zen` folders). Stacks are no longer stored in
`.zen` folders though, only the locally active profile and locally active stack
are maintained there. This is an example of what the local `.zen/config.yaml`
file looks like:

```yaml
active_profile_name: stefan
active_stack_name: default
```

* the Repository acts as an overlay that sits between the users and the
GlobalConfig. The _active profile_ and _active stack_ are
objects that can be configured both globally and at repository level
(when a repository root is initialized) and also given that the Repository
local configuration takes precedence over the global configuration. This
is made evident by the introduction of the `BaseGlobalConfiguration` abstract
class, which is implemented by both GlobalConfig and Repository.
 
* moved `version` from repository to global configuration, because it is
an attribute that needs to be addressed globally e.g. when migrating the
global configuration to a new version. Implemented rudimentary schema migration
and validation for the global configuration.

* since local .zen repositories are no longer mandatory, this PR also replaces 
.zen folders as reference points for building container images and loading source
path classes with dynamic detection of root python paths

* as a migration path from legacy .zen folders where configured stacks are
still stored, the user has the option to create a profile that points to the .zen
folder location and use it in its stead

* this PR implements new Profile related CLI commands and updates
the existing CLI commands related to stacks and stack components to
reflect the newly added Profile component

* the default stack and stack components are renamed to `default`

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

schustmi commented on src/zenml/integrations/kubeflow/orchestrators/kubeflow_dag_runner.py:
```diff
@@ -350,12 +348,7 @@ def _construct_pipeline_graph(
-            main_module = source_utils.get_module_source_from_file_path(
-                os.path.abspath(main_module_file)
-            )
+            main_module = sys.modules["__main__"].__name__
```
I think the name of the main module will always be `__main__`. Let me explain why this matters when passing it to kubeflow: Inside the kubeflow entrypoint, we try to import the user main file in order to also import any materializers or required custom code that got imported when the pipeline is run locally. When we're passing `__main__`, the entrypoint won't know which file to import (and in case the step is in the same file, won't be able to find the step implementation either). Maybe there is some detail that I'm missing as I haven't quite finished the review, but it seems like this should fail for many cases (same with the step operators entrypoint)

stefannica replied:
This is something that Kubeflow integration test cases also caught. I switched everything back to how it was (more or less), but I still allow code to run outside of a .zen repository, in which case the "root" is considered to be the folder where
the main module file is located.

schustmi commented on src/zenml/utils/source_utils.py:
```diff
@@ -277,27 +280,7 @@ def resolve_class(class_: Type[Any]) -> str:
-    # module_source = module_source.replace("\\", ".")
-
-    return module_source + "." + class_.__name__
+    return initial_source
```
I think this will "break" the post-execution workflow in the following way:
For each output artifact, we store the resolved string (e.g. `some.module.SomeClass`) of the materializer and datatype inside the metadata store (see `zenml/steps/utils.py` method `resolve_output_artifact`). When trying to read an artifact in post-execution, we load those classes to read the artifact. 
With the source resolved like this, not only would post-execution only work from the exact same directory as you ran the pipeline, but any materializer/custom data type defined in the main module will fail to import.

stefannica replied:
Same as above, I restored the original code, with a code path that allows users to run zenml outside of a .zen repository. When no .zen repo is found, the "root" path against which the module source is computed is the folder where the main module file is located

schustmi reviewed (changes requested):
Really really nice! I haven't finished yet (still missing the two large beasts GlobalConfig and Repository as well as the tests), but already found some potential issues I thought were worth mentioning early.

htahir1 commented:
I have not looked at the code but took on the persona of the user to see how this felt. I have the following feedback:

* The first time I did this, I had no idea what a profile really was -> It would be nice to link to the documentation after I do a zenml init. I also tried `zenml profile explain` but it was not there
* I did not understand what the difference was between 👑and👉 without any clear guidelines
* From your notes: "as a migration path from legacy .zen folders where configured stacks are still stored, the user has the option to create a profile that points to the .zen folder location and use it in its stead" -> I am unsure how to do this.
* As there are no docs changes here, I would assume we would do the docs changes at a later point? Would be nice to get this whole profile stuff in the docs ASAP because it seems to affect many different points of the docs. I am making a seperate docs update branch as we speak so maybe we can merge that

Generally, my comments are mostly from the perspective of a simple user, and not any sort of opinion of the quality of this PR on the code level. I think whats been done is excellent and I really like this notion of profiles!

stefannica commented on src/zenml/stack_stores/local_stack_store.py:
```diff
@@ -32,37 +31,76 @@
+    @staticmethod
+    def is_valid_url(url: str) -> bool:
+        """Check if the given url is a valid local path."""
+        return url.startswith("file://") or url.startswith("/")
```
Right ! Good catch ! I'll change it to cover windows paths

stefannica commented:
@htahir1 thanks for the feedback, please see inline...

> I have not looked at the code but took on the persona of the user to see how this felt. I have the following feedback:
> 
> * The first time I did this, I had no idea what a profile really was -> It would be nice to link to the documentation after I do a zenml init. I also tried `zenml profile explain` but it was not there

I added the `explain` command and pointed to it in the `zenml init output`:

```
$ zenml init
ZenML repository initialized at /tmp/zenml.
The local active profile was initialized to 'default' and the local active stack to 'default'. This local configuration will only take effect when you're running ZenML from the initialized repository root, or 
from a subdirectory. For more information on profile and stack configuration, please run 'zenml profile explain'.
```

> * I did not understand what the difference was between crownandpoint_right without any clear guidelines

I appended a legend to those tables:

```
$ zenml stack list
Runnning without an active repository root.
Running with active profile: `default` (global)
┏━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━┓
┃ ACTIVE │ STACK NAME │ ARTIFACT_STORE │ METADATA_STORE │ ORCHESTRATOR ┃
┠────────┼────────────┼────────────────┼────────────────┼──────────────┨
┃  👑👉  │ default    │ default        │ default        │ default      ┃
┗━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━┛
👑 = globally active, 👉 = locally active                               
```

> * From your notes: "as a migration path from legacy .zen folders where configured stacks are still stored, the user has the option to create a profile that points to the .zen folder location and use it in its stead" -> I am unsure how to do this.

You'll see a message being printed when you're running this code in a legacy repository. However, that's not very nice. One last point on my list that I didn't get to do yet is to perform the migration transparently, so that users don't need to do anything and don't even need to be aware of it.

> * As there are no docs changes here, I would assume we would do the docs changes at a later point? Would be nice to get this whole profile stuff in the docs ASAP because it seems to affect many different points of the docs. I am making a seperate docs update branch as we speak so maybe we can merge that

I added the Profile to the concepts page. I checked the rest of the docs and there doesn't seem to be any particular need to update anything else. The introduction of Profiles doesn't really affect that much at this point. Profiles will become relevant when we implement the Zen Service and use Profiles that can be configured to use Zen Service as a back-end, but for now, there is not much use for Profiles.

> 
> Generally, my comments are mostly from the perspective of a simple user, and not any sort of opinion of the quality of this PR on the code level. I think whats been done is excellent and I really like this notion of profiles!

stefannica commented:
> * From your notes: "as a migration path from legacy .zen folders where configured stacks are still stored, the user has the option to create a profile that points to the .zen folder location and use it in its stead" -> I am unsure how to do this.

I implemented an automated migration of legacy repositories, so this doesn't need to rely on explicit user interaction anymore:
![Screenshot from 2022-03-23 11-13-49](https://user-images.githubusercontent.com/3963946/159676947-f3e65f2b-dbdc-4045-9222-79242331f444.png)

schustmi reviewed (changes requested):
Amazing PR! I took a look and everything and only have a few minor comments, although I certainly missed some things due to the sheer amount of changes here. The class hierarchy and many config directories were a little confusing to me initially, but after this second review I think I have at least an elementary idea of what's happening and really like the your implementation.

stefannica commented on tests/unit/orchestrators/test_utils.py:
```diff
@@ -33,7 +34,7 @@ def test_get_cache_status_raises_no_error_when_none_passed():
 
 
-def test_get_cache_status_works_when_running_pipeline_twice(clean_repo):
+def test_get_cache_status_works_when_running_pipeline_twice():
```
Sure. I created a new, `autouse` module-level fixture called `base_profile` that creates a separate profile for every unit test module. This means that `clean_repo` is no longer required, because that guarantees separation of stacks at module level. The only tests that still need a `clean_repo` (meaning a completely new global configuration and set of profiles after my PR) are the repository and global configuration tests. Even the examples benefit from the same profile separation. 

The bonus is that profiles are now also actively being tested as part of the test setup itself.

schustmi replied:
Ah that clears at least one of my concerns, I somehow mistakenly remembered that the `base_profile` was a session-scoped fixture similar to `base_repo`. I'm guessing we could still run into trouble if the same testing module ran a pipeline with the same name and step names, in which case the caching results might be wrong depending on the order of the tests? Probably not a real concern and seems to not happen in any tests, just something to keep in mind for future tests. I see the risk especially in modules that use our fixtures to get pipelines/steps, in which case the overlap might happen more frequently

stefannica commented:
@schustmi I addressed all your comments in the last commit.

jwwwb commented on src/zenml/cli/stack_components.py:
```diff
@@ -69,47 +76,93 @@ def generate_stack_component_get_command(
+        "-g",
+        "global_profile",
+        is_flag=True,
+        help=f"Show the global active {singular_display_name}.",
```
"describe the stack component from global" maybe?

stefannica replied:
I removed this entirely.

jwwwb commented on src/zenml/cli/stack_components.py:
```diff
@@ -38,13 +38,16 @@ def _component_display_name(
 def _get_stack_component(
     component_type: StackComponentType,
     component_name: Optional[str] = None,
+    active_stack_name: Optional[str] = None,
```
Little bit confused by this variable name. it's just which stack to get the component_type from?
I would suggest renaming to `stack_name` and furthermore documenting + enforcing that only one of `{component_name, stack_name}` may be not None, because right now the behavior is unclear, obviously one of them has to take preference, but it's better to make this explicit.

stefannica replied:
Same here, I removed it completely as part of the simplification of CLI commands to avoid the global/local complications.

jwwwb commented on src/zenml/cli/stack_components.py:
```diff
@@ -122,14 +175,18 @@ def describe_stack_component_command(name: Optional[str]) -> None:
+        if active_stack_name:
+            try:
+                active_component_name = (
+                    repo.get_stack(active_stack_name)
```
Does `repo.active_stack` no longer work? How about making it work again by putting all of this logic from here into that property?

stefannica replied:
Same here, I removed it. But yes, `repo.active_stack` still works fine.

jwwwb commented on src/zenml/cli/stack_components.py:
```diff
@@ -145,21 +202,43 @@ def generate_stack_component_list_command(
+        local_stack = repo.active_stack_name
+        global_stack = repo.active_profile.active_stack or ""
+
+        global_active_component_name = None
```
Referring to my other comment about what to show the user. showing the user both the globally active and the locally active version of everything introduces so much extra code overhead in many places. if we instead just show the local information, except when either currently outside of a repo or supplying a --global flag, I feel like this would simplify a lot of implementations, and doesn't deprive the user of any valuable information (if they reallly need to know both, they just issue the command twice).

stefannica replied:
I completely agree and now that I've gone through the exercise of removing the global/local distinction wherever it didn't make sense anymore, I'm flabbergasted by the amount of useless code that I took out.

jwwwb reviewed (changes requested):
I have not finished the whole change review, but I think it makes sense to leave this intermediate feedback already. My main suggestion would be to not show local and global information at the same time. That would reduce code amount significantly, and less information for the user is probably an improvement in understanding.

stefannica commented on src/zenml/cli/config.py:
```diff
@@ -82,3 +90,376 @@ def set_logging_verbosity(verbosity: str) -> None:
+def explain_profile() -> None:
+    """Explains the concept of ZenML profiles."""
+
+    console.print(
```
There actually is, thank you for the suggestion :)

stefannica commented on src/zenml/cli/config.py:
```diff
@@ -82,3 +90,376 @@ def set_logging_verbosity(verbosity: str) -> None:
+┏━━━━━━━━┯━━━━━━━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━┓
+┃ ACTIVE │ PROFILE NAME │ STORE TYPE │ URL                                                │ GLOBAL ACTIVE STACK │ LOCAL ACTIVE STACK ┃
+┠────────┼──────────────┼────────────┼────────────────────────────────────────────────────┼─────────────────────┼────────────────────┨
+┃  👑👉  │ default      │ local      │ file:///home/stefan/.config/zenml/profiles/default │ default             │ default            ┃
```
I squeezed the console examples so they fit into 80 characters now.

stefannica commented on src/zenml/cli/stack.py:
```diff
@@ -124,20 +126,26 @@ def register_stack(
+        if stack_name == global_stack:
+            active_str += ":crown:"
+        if stack_name == repo.active_stack_name:
+            active_str += ":point_right:"
```
Good points. I removed the global/local differentiation here:

```
/tmp/zenml$ zenml stack list
Running with active profile: 'zenml' (local)
┏━━━━━━━━┯━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━┓
┃ ACTIVE │ STACK NAME │ ARTIFACT_STORE │ METADATA_STORE │ ORCHESTRATOR ┃
┠────────┼────────────┼────────────────┼────────────────┼──────────────┨
┃   👉   │ default    │ default        │ default        │ default      ┃
┃        │ local      │ default        │ default        │ default      ┃
┗━━━━━━━━┷━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━┛
```

stefannica commented on src/zenml/cli/utils.py:
```diff
@@ -217,6 +249,71 @@ def print_stack_component_configuration(
+    if repo.root:
+        scope = "local"
+    else:
+        scope = "global"
```
I agree. I fixed this and any other occurrences of this pattern wherever I could find them.

stefannica commented on src/zenml/cli/utils.py:
```diff
@@ -217,6 +249,71 @@ def print_stack_component_configuration(
+    items = profile.dict().items()
+    for item in items:
+        rich_table.add_row(*[str(elem) for elem in item])
+    rich_table.add_row("LOCAL ACTIVE STACK", local_stack)
```
It turns out that active_stack_name was printed twice, because it's already part of the profile model, so I removed this one.

stefannica commented on src/zenml/config/global_config.py:
```diff
@@ -13,19 +13,241 @@
+    def __init__(
+        self, config: Optional["GlobalConfig"] = None, **kwargs: Any
+    ) -> None:
+        """Initializes a GlobalConfig object using values from the config file.
```
Ouch, you found a copy-paste mishap. Thanks for catching that !

stefannica commented:
> I have not finished the whole change review, but I think it makes sense to leave this intermediate feedback already. My main suggestion would be to not show local and global information at the same time. That would reduce code amount significantly, and less information for the user is probably an improvement in understanding.

Your feedback was priceless. I removed the global/local confusion from all the CLI commands where it no longer made sense (the only exceptions were the `zenml profile set` and `zenml stack set` commands where choosing the scope is important as a shortcut when the active profile/stack cannot be deleted. Thanks to you, not only does the CLI look cleaner and easier to understand now, but there was also a relatively large amount of code removed that is no longer part of the PR.

jwwwb commented on src/zenml/stack_stores/sql_stack_store.py:
```diff
@@ -56,44 +59,98 @@ class ZenStackDefinition(SQLModel, table=True):
+        """
+        if not SqlStackStore.is_valid_url(url):
+            raise ValueError(f"Invalid URL for SQL store: {url}")
+        if not url.startswith("sqlite:///"):
```
I'm not sure I understand the difference between this and the check 2 lines above. how can it be a valid url but not start with this string? Furthermore, why return None in this case?

stefannica replied:
The SqlStackStore check is borrowed from SQLAlchemy: it checks that it is a valid SQLAlchemy URL, which can take many forms for the many drivers and SQL databases that it supports. This function only applies to the SQLite case that stores the DB information inside a local file. All other cases are non-local and don't have an associated local path.

jwwwb commented on src/zenml/config/global_config.py:
```diff
@@ -106,20 +430,254 @@ def _read_config(self) -> Dict[str, Any]:
+        """
+        return os.path.join(config_path or self._config_path, "config.yaml")
+
+    def copy_config_with_active_profile(
```
I like the idea of this function, but I feel like it could be made simpler by just always storing the contents of the profile into a new YAML_STORAGE based one, instead of copying along relative paths from who knows where and adjusting the urls to those.

jwwwb commented on tests/unit/cli/test_base.py:
```diff
@@ -17,11 +17,11 @@
 
 from zenml.cli.base import init
-from zenml.repository import LOCAL_CONFIG_DIRECTORY_NAME
+from zenml.constants import REPOSITORY_DIRECTORY_NAME
```
I'm cognizant of the irony of me seeing this change and using it to suggest the following:

Initially the intent of this PR was to remove the "Repository" class entirely and handle all of its tasks on a profile or global config level. Although we then realized that we can't avoid keeping some configuration inside of every project's folder.... I wonder if it might not still be helpful to remove the name "Repository" at least. Really, whatever this directory still is, I don't think it's a repository. Maybe rename the Repository to LocalConfig, to contrast it with GlobalConfig? Overall, I think the naming of the classes in this PR could benefit from a bit more clarity, just being more descriptive of what they actually do. 

Currently:
* GlobalConfig
* ConfigProfile
* Repository

Suggestion
* GlobalConfiguration
* ProfileConfiguration
* LocalConfiguration? 

Maybe something like that. Unless I'm misunderstanding what these objects do

stefannica replied:
We can't rename Repositoy for two main reasons:
1. Repository has been part of the public API and is a very central concept that is already heavily in use in code by our users everywhere. Sure, we can still have something like a `Repository = LocalConfiguration` dummy placeholder for it (but see my other reason).
2. Repository is more than just configuration. It is a facade for many different things, such as managing stacks, retrieving pipeline run results and I imagine it will extend to cover all other information involved in the Zen Service communication (Users, Teams, maybe even Secrets ?).

I do agree in principle that `Repository` is not a good name for it anymore and that `LocalConfiguration` is a set of concerns that can be detached from it and moved elsewhere. That's what inspired me to create that `BaseGlobalConfiguration` abstract class that is implemented both by GlobalConfig and Repository, but that is a more convoluted change that should be done separately, if not out of concern for the size of this PR, then at least for time considerations (this PR needs to go in the next release).

jwwwb commented on src/zenml/config/global_config.py:
```diff
@@ -13,19 +13,244 @@
 
-class GlobalConfig(BaseModel):
+
+class ConfigProfile(BaseModel):
```
This is already a huge file, maybe this class could be moved to its own file? (and perhaps renamed, as I suggested below)

jwwwb reviewed (changes requested):
Great idea overall, I worry that some bits are more complex than they need to be. This PR has many lines of code, and that's all lines to be maintained in the future, for contributors to have to understand, and even for end-users to have to understand if they want to look under the hood. I made a few more suggestions that I think would improve the clarity of reading the code, but other those, I've (finally) finished reading and I like it.

stefannica commented:
@jwwwb Please check my latest additions to this PR. I made a few renames, as you suggested, and moved the ConfigProfile (now ProfileConfiguration) to a separate file.

jwwwb reviewed (approved):
Thanks for those changes, makes things clearer!

strickvl commented:
Now that all the initial stacks after `zenml init` are called `default`, this probably means that we'll have to change all the places in our documentation that refer to `local_stack` or `local_orchestrator` etc.
